packages/components/src/custom-select-control/test/index.tsx

```diff
@@ -689,3 +690,127 @@ describe.each( [
 		} );
 	} );
 } );
+
+describe( 'Type checking', () => {
+	// eslint-disable-next-line jest/expect-expect
+	it( 'should infer the value type from available `options`, but not the `value` or `onChange` prop', () => {
+		const options = [
+			{
+				key: 'narrow',
+				name: 'Narrow',
+			},
+			{
+				key: 'value',
+				name: 'Value',
+			},
+		];
+		const optionsReadOnly = [
+			{
+				key: 'narrow',
+				name: 'Narrow',
+			},
+			{
+				key: 'value',
+				name: 'Value',
+			},
+		] as const;
+
+		const onChange: (
+			changeObject: CustomSelectChangeObject< CustomSelectOption >
+		) => void = () => {};
+		const onChangeWithFoo: (
+			changeObject: CustomSelectChangeObject<
+				CustomSelectOption & {
+					foo: string;
+				}
+			>
+		) => void = () => {};
```

Comment:
Ideally we would write these tests without having to import the `CustomSelectChangeObject` and `CustomSelectOption` types directly. Consumers outside of this package don't have access to these internal types.

```suggestion
		const onChange: React.ComponentProps<
			typeof UncontrolledCustomSelectControl
		>[ 'onChange' ] = () => {};
```

And for the `onChangeWithFoo` cases, I think we could simplify it into a more realistic example and just inline it:

```tsx
		<UncontrolledCustomSelectControl
			label="Label"
			options={ options }
			value={ {
				key: 'narrow',
				name: 'Narrow',
			} }
			// @ts-expect-error: the option type should not be inferred from `onChange`
			onChange={ ( obj ) => obj.foo }
		/>;
```

Reply:
Updated.

As for the `onChangeWithFoo` cases, what you provided above is not correct. We need to test if the `NoInfer` in the `onChange` method works as expected, so we have to define the type of the `onChange` prop explicitly.

It should not be a realistic use case for sure, but we still need to test it. So I added some comments to explain it.

Reply:
Ah, you're right. Thanks for the simplification and code comment though, it reads a lot better 👍